### ISSUE TITLE
[Feature Request]: Align Auto-complete brackets with the native Obsidian implementation

### ISSUE THREAD

**Issue by infinityiowastaken** [joeyuping/quick_latex_obsidian#15]
This is more of an optional request as opposed to an 'issue', but would it be possible to make the auto-complete components work more closely to that of brackets outside of LaTeX blocks?

This would require implementing the following rule for bracket-style characters: `{ }` / `[ ]` / `( )`:

If there is a significant character (all characters that are not `$`, newline, space or tab?) to the right of the cursor when the opening character is pressed, do not auto-close the bracket:

```c
// default behaviour, cursor is at | symbol
$|a$
// '(' input
$(|)a$

// desired behaviour
$|a$
// '(' input
$(|a$
```

Thank you!

**Comment by joeyuping** [comment]
Thank you! indeed! I did notice about this. Often we may type an expression, after that realizing that it needs a bracket and goes back to add the bracket. The auto-complete bracket function becomes a hindrance in such use case. I have been thinking whether to adjust it.

What I did though, was to add another function (see Readme point 6), where one can select the expression that needs to be enclosed with brackets then press the corresponding open bracket ( i.e.  "{", "[", "("  ) to automatically enclose the expression in brackets. I wonder if this may be a better way to cover the use case as mentioned above? In that case, would it then be desirable to keep the auto-complete bracket simple and without exception, less some users may feel that the auto-complete bracket function is buggy, working in some cases and not in some others?

What do you think?

Thank you so much for bringing this up!

**Comment by infinityiowastaken** [comment]
I think it's entirely preference so not necessarily something that needs to be fixed, but I realised that I was being caught out occasionally by the fact that (as far as I can tell from my own testing) Obsidian natively implements this slightly stranger version as described above, and as such I think I was struggling to quickly swap between the two variation, which led to it feeling slightly strange when I was swapping between LaTeX and non-LaTeX editing. However, in the end it is all down to personal preference

Thanks!

**Comment by joeyuping** [comment]
Thankyou! I agree with you. Indeed it would work better that the same bracket function works the same way through out obsidian. 

I tried to implement this by only firing the auto-complete bracket function when the following character is $, space, tab or nothing.

After some preliminary testing, it does seem more intuitive ~ 

Please check out version 1.5.3, and help me test if there is any other situation that may seem awkward when using!  @infinityiowastaken

**Comment by infinityiowastaken** [comment]
This does seem to be significantly more easy to use, thank you! My only remaining note is that obsidian will not autoclose if the cursor is immediately before a close bracket, as far as I can tell? (`a|)` leads to `a(|)`)

**Comment by richukuttan** [comment]
I'd like to request one quick change if possible. Could you also autocomplete the bracket if the next two characters are `\\`? This arises in aligns, where I'm editing at the end of the line. Logically, this is a form of whitespace, so the autocomplete should work.

**Comment by joeyuping** [comment]
@richukuttan @infinityiowastaken thank you! have noted the above! will add them into the next version!

**Comment by richukuttan** [comment]
Thanks, one more edge case discovered - If I simply type `(` between `$$$$`, it doesn't autocomplete because the next character is `$`.

**Comment by joeyuping** [comment]
Thanks all!!

I have resolved the above in v1.5.5, please take a look~

@infinityiowastaken hi! I could not reproduce the case you mentioned though... in normal obsidian usage, I see that obsidian autocloses bracket when immediately before a close bracket. So I kept it as it is.
